stdio/format.c

```diff
@@ -81,303 +118,624 @@ static inline uint64_t format_u64FromDouble(double d)
+	int res = format_bigdoubleInit(result, 1024);
+	bignum_zero(&result->num);
+	if (isSubnormalOrZero != 0 && mantissa == 0) {
+		/* val == 0 */
+		result->firstIntegerBit = 0;
+		return 0;
+	}
+	/* val != 0 */
+	mantissa |= (isSubnormalOrZero != 0 ? 0LLU : 1LLU) << DOUBLE_EXP_SHIFT;
+	result->num.data[0] = mantissa & BIGNUM_DATA_MASK;
+	mantissa >>= BIGNUM_DATA_BITS;
+	res = bignum_push(&result->num, (bignum_data_t)mantissa);
+	if (res != 0) {
+		format_bigdoubleDestroy(result);
+		return res;
+	}
+	if (isSubnormalOrZero != 0) {
+		exp = -1022;
+	}
+	/* Check if there is no fractional part */
+	if (DOUBLE_EXP_SHIFT < exp) {
+		result->firstIntegerBit = 0;
+		res = bignum_shiftl(&result->num, exp - DOUBLE_EXP_SHIFT);
+	}
+	else {
+		result->firstIntegerBit = DOUBLE_EXP_SHIFT - exp;
+	}
+	return res;
 }
 
 
-static inline uint32_t format_fracToU32(float frac, int float_frac_len, float *overflow)
+static void format_printBuffer(void *ctx, feedfunc feed, uint32_t flags, int minFieldWidth, const char *start, const char *end, char sign)
 {
-	const uint32_t s_powers10[] = { 1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000, 1000000000, };
-	const uint32_t mult = s_powers10[float_frac_len];
+	const int digits_cnt = start > end ? start - end : end - start;
+	int pad_len = minFieldWidth - digits_cnt - (sign ? 1 : 0);
+	/* If FLAG_ZERO and FLAG_MINUS both appear, then FLAG_ZERO is ignored */
+	if ((flags & FLAG_MINUS) != 0) {
+		flags &= ~FLAG_ZERO;
+	}
+
+	/* pad, if needed */
+	if (pad_len > 0 && (flags & FLAG_MINUS) == 0 && (flags & FLAG_ZERO) == 0) {
+		while (pad_len-- > 0) {
+			feed(ctx, ' ');
+		}
+	}
 
-	frac *= (float)mult;
+	if (sign != 0) {
+		feed(ctx, sign);
+	}
 
-	/* Ensure proper rounding */
-	frac += 0.5f;
-	uint32_t ret = (uint32_t)(int32_t)frac;
+	/* pad, if needed */
+	if (pad_len > 0 && (flags & FLAG_MINUS) == 0 && (flags & FLAG_ZERO) != 0) {
+		while (pad_len-- > 0) {
+			feed(ctx, '0');
+		}
+	}
 
-	if (ret >= mult) {
-		*overflow += 1;
-		ret -= mult;
+	/* copy */
+	if (start > end) {
+		while ((--start) >= end) {
+			feed(ctx, *start);
+		}
+	}
+	else {
+		while (start < end) {
+			feed(ctx, *start++);
+		}
+	}
+
+	/* pad, if needed */
+	if (pad_len > 0 && (flags & FLAG_MINUS) != 0) {
+		while (pad_len-- > 0) {
+			feed(ctx, ' ');
+		}
 	}
-	return ret;
 }
 
 
-static void format_sprintf_num(void *ctx, feedfunc feed, uint64_t num64, uint32_t flags, int min_number_len, int float_frac_len)
+static inline char format_chooseSign(int sign /* 0 means + */, uint32_t flags)
 {
-	const char *digits = (flags & FLAG_LARGE_DIGITS) ? "0123456789ABCDEF" : "0123456789abcdef",
-			*prefix = (flags & FLAG_LARGE_DIGITS) ? "X0" : "x0";
-	char tmp_buf[32];
-	char sign = 0;
-	char *tmp = tmp_buf;
+	if (sign != 0) {
+		return '-';
+	}
+	else if ((flags & FLAG_PLUS) != 0) {
+		return '+';
+	}
+	else if ((flags & FLAG_SPACE) != 0) { /* If FLAG_PLUS and FLAG_SPACE both appear, then ignore SPACE */
+		return ' ';
+	}
+	else {
+		return 0;
```

Comment:
MISRA - unwanted cast int -> char
```suggestion
		return '\0';
```